sklearn/metrics/pairwise.py

```diff
@@ -1130,6 +1131,185 @@ def _pairwise_callable(X, Y, metric, **kwds):
+
+    Y : array [n_samples_b, n_features], optional
+        An optional second feature array. Only allowed if
+        metric != "precomputed".
+
+    metric : string, or callable
+        The metric to use when calculating distance between instances in a
+        feature array.
+
+    n_jobs : int
+        The number of jobs to use for the computation.
+
+    block_size : int, default=64
+        The maximum number of mebibytes (MiB) of memory per job (see``n_jobs``)
+        to use at a time for calculating pairwise distances.
+
+    block_n_rows : int
+        Number of rows to be computed for each block.
+
+    `**kwds` : optional keyword parameters
+        Any further parameters are passed directly to the distance function.
+
+    Returns
+    -------
+    D : generator of blocks based on the ``block_size`` parameter.
+
+    """
+    if metric != 'precomputed' and Y is None:
+        Y = X
+    n_samples = X.shape[0]
+    for start in range(0, n_samples, block_n_rows):
+        # get distances from block to every other sample
+        stop = min(start + block_n_rows, X.shape[0])
+        yield pairwise_distances(X[start:stop], Y, metric, n_jobs, **kwds)
+
+
+def _generate_pairwise_distances_reduce(X, Y=None, metric='euclidean',
+                                        n_jobs=1, reduce_func=None,
+                                        block_size=DEFAULT_BLOCK_SIZE,
+                                        block_n_rows=1, **kwds):
+    if metric != 'precomputed' and Y is None:
+        Y = X
+    n_samples = X.shape[0]
+    if reduce_func:
+        for start in range(0, n_samples, block_n_rows):
+            # get distances from block to every other sample
+            stop = min(start + block_n_rows, X.shape[0])
+            dist = pairwise_distances(X[start:stop], Y, metric, n_jobs, **kwds)
+            yield reduce_func(dist, stop-start, *kwds)
+
+
+def pairwise_distances_reduce(X, Y=None, metric='euclidean', n_jobs=1,
+                              reduce_func=None, block_size=DEFAULT_BLOCK_SIZE,
+                              **kwds):
+    if (metric not in _VALID_METRICS and
+            not callable(metric) and metric != "precomputed"):
+        raise ValueError("Unknown metric %s. "
+                         "Valid metrics are %s, or 'precomputed', or a "
+                         "callable" % (metric, _VALID_METRICS))
+
+    n_samples = X.shape[0]
+    block_n_rows = block_size * (2 ** 20) // (BYTES_PER_FLOAT * n_samples)
+    if block_n_rows > n_samples:
+        block_n_rows = min(block_n_rows, n_samples)
+    if block_n_rows < 1:
+        min_block_mib = np.ceil(n_samples * BYTES_PER_FLOAT * 2 ** -20)
+        raise ValueError('block_size should be at least n_samples * %d bytes '
+                         '= %.0f MiB, got %r' % (BYTES_PER_FLOAT,
+                                                 min_block_mib, block_size))
+
+    return _generate_pairwise_distances_reduce(X, Y, metric=metric,
```

Comment:
I don't think we should be making changes according to the value of `return_distances` in `pairwise_distances_reduce` as it would not make it very general. Returning a generator allows us to handle that case gracefully in `neighbors/base.py`?